tests/server.rs

```diff
@@ -1482,6 +1485,303 @@ async fn http_connect_new() {
+        assert_eq!(response.status(), StatusCode::OK);
+
+        let mut body = response.into_body();
+        let bytes = body.data().await.unwrap().unwrap();
+        assert_eq!(&bytes[..], b"Bread?");
+        let _ = body.flow_control().release_capacity(bytes.len());
+
+        (body, send_stream)
+    }
+
+    tokio::spawn(async move {
+        let (mut recv_stream, mut send_stream) = connect_and_recv_bread(&mut h2).await;
+
+        send_stream.send_data("Baguette!".into(), true).unwrap();
+
+        assert!(recv_stream.data().await.unwrap().unwrap().is_empty());
+    });
+
+    let svc = service_fn(move |req: Request<Body>| {
+        let on_upgrade = hyper::upgrade::on(req);
+
+        tokio::spawn(async move {
+            let mut upgraded = on_upgrade.await.expect("on_upgrade");
+            upgraded.write_all(b"Bread?").await.unwrap();
+
+            let mut vec = vec![];
+            upgraded.read_to_end(&mut vec).await.unwrap();
+            assert_eq!(s(&vec), "Baguette!");
+
+            upgraded.shutdown().await.unwrap();
+        });
+
+        future::ok::<_, hyper::Error>(
+            Response::builder()
+                .status(200)
+                .body(hyper::Body::empty())
+                .unwrap(),
+        )
+    });
+
+    let (socket, _) = listener.accept().await.unwrap();
+    Http::new()
+        .http2_only(true)
+        .serve_connection(socket, svc)
+        .with_upgrades()
+        .await
+        .unwrap();
+}
+
+#[tokio::test]
+async fn h2_connect_multiplex() {
+    use tokio::io::{AsyncReadExt, AsyncWriteExt};
+
+    let _ = pretty_env_logger::try_init();
+    let listener = tcp_bind(&"127.0.0.1:0".parse().unwrap()).unwrap();
+    let addr = listener.local_addr().unwrap();
+    let conn = connect_async(addr).await;
+
+    let (h2, connection) = h2::client::handshake(conn).await.unwrap();
+    tokio::spawn(async move {
+        connection.await.unwrap();
+    });
+    let mut h2 = h2.ready().await.unwrap();
+
+    async fn connect_and_recv_bread(
+        h2: &mut SendRequest<Bytes>,
+    ) -> (RecvStream, SendStream<Bytes>) {
+        let request = Request::connect("localhost").body(()).unwrap();
+        let (response, send_stream) = h2.send_request(request, false).unwrap();
+        let response = response.await.unwrap();
+        assert_eq!(response.status(), StatusCode::OK);
+
+        let mut body = response.into_body();
+        let bytes = body.data().await.unwrap().unwrap();
+        assert_eq!(&bytes[..], b"Bread?");
+        let _ = body.flow_control().release_capacity(bytes.len());
+
+        (body, send_stream)
+    }
+
+    tokio::spawn(async move {
+        let (mut recv_stream_1, mut send_stream_1) = connect_and_recv_bread(&mut h2).await;
+
+        let (_, mut send_stream_2) = connect_and_recv_bread(&mut h2).await;
+
+        send_stream_1.send_data("Baguette!".into(), true).unwrap();
+
+        send_stream_2.send_reset(h2::Reason::PROTOCOL_ERROR);
```

Comment:
I would like more streams :) Let's say 50 transfering data and 50 error.

Reply:
Also, in case of error, is it propagated to the other peer's Upgraded "socket"?

Reply:
let's cover that in the test too (if not yet, sorry, didn't have a chance to give it a full round of review yet)

Reply:
Added some assertion to check the error kind and the error payload, with 80 clients stopping at 4 different points.